modules/videoio/src/cap_v4l.cpp

```diff
@@ -1880,7 +1830,11 @@ bool CvCaptureCAM_V4L::grabFrame()
 
 IplImage* CvCaptureCAM_V4L::retrieveFrame(int)
 {
-    return icvRetrieveFrameCAM_V4L( this, 0 );
+    IplImage *image = icvRetrieveFrameCAM_V4L(this, 0);
+    //Revert buffer to the queue
+    if (-1 == ioctl(deviceHandle, VIDIOC_QBUF, &buffers[bufferIndex].buffer))
+        perror("VIDIOC_QBUF");
```

Comment:
Not sure that we really need this.

`.retrieve()` can be called several times (even with the same "index" parameter).

Next `.grab()` call should "return back" this buffer.

Reply:
We need to put buffer back to the queue as fast as we can. `Index` parameter is a channel `VideoCapture::retrieve(OutputArray image, int channel)` that has no sense for v4l capture, because changing channel lead to stop streaming. Ok we can call `.retrieve()` several times, I can fixed by send old `IplImage *image`, is it ok?

Reply:
Also `.retrieve()` may not be called at all after `.grab()`.

Reply:
Done